### DISCUSSION TITLE
Confusing "When to use" in readme (micromark vs remark)

### DISCUSSION THREAD

**Discussion by chrisjsewell** [micromark/micromark#110] [Q&A] [answered]
Heya, here

<img width="668" alt="image" src="https://user-images.githubusercontent.com/2997570/156583856-c758de87-5e9e-4147-acb8-07488b79c769.png">

and here (in Comparison):

<img width="858" alt="image" src="https://user-images.githubusercontent.com/2997570/156586321-dc0f1186-f858-4f77-a973-e54b3c70bf1e.png">

I assume the "complex things"  / "fine grained" control are meant to reference remark, rather than saying you should use micromark for both 😬 ?

**Comment by ChristianMurphy** [comment, marked as answer]
Micromark offers two things:
1. a low level parser
2. a minimally configurable markdown to html compiler

---

> If you want to do really complex things with markdown

and

> If you’re looking for fine grained control, use micromark

refer to the former, the low level parser

---

> If you just want to turn markdown into HTML (with maybe a few extensions)

and

> If you just want HTML from markdown, use micromark

refers to the latter, the minimally configurable markdown to html compiler

**Reply by chrisjsewell** [reply]
Thanks,

> Micromark offers two things

I think this should be explained then in the readme because, IMO, the "when to use" currently reads as "use it for everything, simple and complex", which is not very helpful 😬
Conversely, when would you say it should not be used?

Similarly, if micromark can be used for both fine-grained and simple use cases, why would you use remark?

**Reply by chrisjsewell** [reply]
In the remark readme it also says:

<img width="805" alt="image" src="https://user-images.githubusercontent.com/2997570/156591639-a4a51984-89fb-4677-9ce3-8f1920faeceb.png">

i.e. micromark for simple things, remark for complex things,
which again is a bit cognitively dissonant

**Reply by ChristianMurphy** [reply]
> Similarly, if micromark can be used for both fine-grained and simple use cases, why would you use remark?
>
> i.e. micromark for simple things, remark for complex things, which again is a bit cognitively dissonant

it is not dissonant.
micromark lives on the edges of very simple and very complex.
remark lives in the large and happy medium between, and has a large ecosystem of plugins supporting this.

> I think this should be explained then in the readme because, IMO, the "when to use" currently reads as "use it for everything, simple and complex", which is not very helpful 

Pull requests improving the documentation are welcome!

**Reply by chrisjsewell** [reply]
Cheers, do you have any "examples" of something that would probably be too complex for remark, and instead you would use micromark.
I'm 99% sure that my use case is suited to remark, but wanted to check that I was not missing anything by prematurely "disregarding" micromark

**Reply by ChristianMurphy** [reply]
> do you have any "examples" of something that would probably be too complex for remark, and instead you would use micromark.

Sure, a recent example is: https://github.com/mdx-js/mdx/issues/1953
Adding a new syntax feature like `{# heading-id }` needs to happen in micromark, because it potentially overlaps with other syntax in markdown/mdx, and changes the initial AST structure.

Trying to handle this in `remark` would happen too late, in this case, because it already conflicted with a JSX syntax extension and caused an error before the remark plugin was even run.

That said, extending markdown syntax should be done with caution, as noted in https://github.com/micromark/micromark#extending-markdown alternatives, like working in `remark` are often preferable.

https://unifiedjs.com/learn has some guides on working with abstract syntax trees (like mdast/remark) which can help round out the understand of what's possible with ASTs a bit more.

**Reply by chrisjsewell** [reply]
That's great, thanks again. I'll do some more reading before I pester you again 😄
